File: frontend/src/types/workflow.ts
export interface WorkflowNode {
  id: string;
  type: "agent" | "condition" | "start" | "end";
  position: { x: number; y: number };
  data: {
    label: string;
    agentId?: string;
    agentName?: string;
    agentRole?: string;
    agentModel?: string;
    condition?: string;
    [key: string]: unknown;
  };
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  label?: string;
  animated?: boolean;
}

export interface GraphDefinition {
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
}

export interface Workflow {
  id: string;
  name: string;
  description: string;
  is_template: boolean;
  template_id: string | null;
  graph_definition: GraphDefinition;
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  is_template?: boolean;
  template_id?: string | null;
  graph_definition?: GraphDefinition;
}

export interface WorkflowUpdate {
  name?: string;
  description?: string;
  is_template?: boolean;
  graph_definition?: GraphDefinition;
}

File: backend/app/schemas/workflow.py
from datetime import datetime

from pydantic import BaseModel, Field


class WorkflowCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str = ""
    is_template: bool = False
    template_id: str | None = None
    graph_definition: dict = Field(default_factory=dict)


class WorkflowUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    is_template: bool | None = None
    graph_definition: dict | None = None


class WorkflowResponse(BaseModel):
    id: str
    name: str
    description: str
    is_template: bool
    template_id: str | None
    graph_definition: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
